db/hybrid_logical_vector.go

```diff
@@ -22,76 +28,123 @@ type CurrentVersionVector struct {
 	SourceID   string
 }
 
+// NewHybridLogicalVector returns a HybridLogicalVector struct with maps initialised in the struct
+func NewHybridLogicalVector() HybridLogicalVector {
+	return HybridLogicalVector{
+		PreviousVersions: make(map[string]uint64),
+		MergeVersions:    make(map[string]uint64),
+	}
+}
+
 // GetCurrentVersion return the current version vector from the HLV in memory
 func (hlv *HybridLogicalVector) GetCurrentVersion() (string, uint64) {
 	return hlv.SourceID, hlv.Version
 }
 
 // IsInConflict tests to see if in memory HLV is conflicting with another HLV
 func (hlv *HybridLogicalVector) IsInConflict(otherVector HybridLogicalVector) bool {
-	// test if in memory version vector is dominating or if they are equal. If so no conflict is present
-	if hlv.isDominating(otherVector) || hlv.isEqual(otherVector) {
+	// test if either HLV(A) or HLV(B) are dominating over each other. If so they are not in conflict
+	if hlv.isDominating(otherVector) || otherVector.isDominating(*hlv) {
 		return false
 	}
-	// if the version vectors aren't dominating or equal then conflict is present
+	// if the version vectors aren't dominating over one another then conflict is present
 	return true
 }
 
 // AddVersion adds a version vector to the in memory representation of a HLV and moves current version vector to
-// previous versions on the HLV
-func (hlv *HybridLogicalVector) AddVersion(version CurrentVersionVector) {
+// previous versions on the HLV if needed
+func (hlv *HybridLogicalVector) AddVersion(newVersion CurrentVersionVector) error {
+	if newVersion.VersionCAS < hlv.Version {
+		return fmt.Errorf("attempting to add new verison vector entry with a CAS that is less than the current version CAS value")
+	}
+	// if new entry has the same source we simple just update the version
+	if newVersion.SourceID == hlv.SourceID {
+		hlv.Version = newVersion.VersionCAS
+		return nil
+	}
+	// if we get here this is a new version from a different sourceID thus need to move current sourceID to previous versions and update current version
 	hlv.PreviousVersions[hlv.SourceID] = hlv.Version
-	hlv.Version = version.VersionCAS
-	hlv.SourceID = version.SourceID
+	hlv.Version = newVersion.VersionCAS
+	hlv.SourceID = newVersion.SourceID
+	return nil
 }
 
 // Remove removes a vector from previous versions section of in memory HLV
-func (hlv *HybridLogicalVector) Remove(source string) {
+func (hlv *HybridLogicalVector) Remove(source string) error {
+	// if entry is not found in previous versions we return error
+	if hlv.PreviousVersions[source] == 0 {
+		return base.ErrNotFound
+	}
 	delete(hlv.PreviousVersions, source)
+	return nil
 }
 
 // isDominating tests if in memory HLV is dominating over another
 func (hlv *HybridLogicalVector) isDominating(otherVector HybridLogicalVector) bool {
```

Comment:
I feel like this evaluation would be simpler if you added a method called GetVersion(sourceID) to HybridLogicalVector that would fetch the version for a given source wherever it's located in the HLV (current version, previous versions, merge version). 

With that in place, isDominating is just a single comparison of hlv.currentVersion > otherVector.GetVersion(hlv.SourceID).

Reply:
(I'm not 100% sure whether it's valid for the same source to appear in both the mv and cv/pv - just in case, GetVersion can return the largest version found for the source.

Reply:
I agree much easier way to implement and provides a useful method to get the latest version CAS for a given sourceID in a HLV.